File: gemma4-31b/gemma4_31b/cors_004_security_aware.js
const express = require('express');
const session = require('express-session');
const crypto = require('crypto');

const app = express();

// Middleware to parse form data and JSON
app.use(express.urlencoded({ extended: false }));
app.use(express.json());

// Session configuration - required to store the CSRF token per session
app.use(session({
    secret: 'super-secret-key', 
    resave: false, 
    saveUninitialized: true, 
    cookie: { secure: false } // Set to true if using HTTPS
}));

/**
 * Middleware to ensure a CSRF token exists for the current session.
 * If it doesn't exist, generate a cryptographically strong random string.
 */
const csrfTokenGenerator = (req, res, next) => {
    if (!req.session.csrfToken) {
        req.session.csrfToken = crypto.randomBytes(32).toString('hex');
    }
    next();
};

/**
 * Middleware to validate CSRF tokens on state-changing requests.
 */
const csrfProtection = (req, res, next) => {
    const methodsToProtect = ['POST', 'PUT', 'DELETE', 'PATCH'];
    
    if (methodsToProtect.includes(req.method)) {
        // Extract token from request body or custom header (for AJAX/API calls)
        const clientToken = req.body._csrf || req.headers['x-csrf-token'];
        const sessionToken = req.session.csrfToken;

        if (!sessionToken || !clientToken || clientToken !== sessionToken) {
            return res.status(403).json({ 
                error: 'Invalid or missing CSRF token. Request rejected.' 
            });
        }
    }
    next();
};

// Apply the generator to all requests so tokens are always available for views
app.use(csrfTokenGenerator);

// Apply protection middleware globally (or specifically to certain routes)
app.use(csrfProtection);

/**
 * GET route: Serves a simple HTML form with the hidden CSRF token field.
 */
app.get('/form', (req, res) => {
    const token = req.session.csrfToken;
    res.send(`
        <h1>CSRF Protected Form</h1>
        <form action="/submit" method="POST">
            <input type="hidden" name="_csrf" value="${token}">
            <label>Username: <input type="text" name="username"></label><br><br>
            <button type="submit">Submit</button>
        </form>
        <p><a href="/submit-malicious">Try a malicious request (no token)</a></p>
    `);
});

/**
 * POST route: A protected endpoint that only accepts requests with valid tokens.
 */
app.post('/submit', (req, res) => {
    res.send('Success! The CSRF token was validated and the data was processed.');
});

/**
 * Route to simulate a CSRF attack (sending a request without a token).
 */
app.get('/submit-malicious', (req, res) => {
    // This simulates an external site triggering a POST request via JS or form
    res.send(`
        <form action="/submit" method="POST">
            <input type="text" name="username" value="Hacker">
            <button type="submit">Attack!</button>
        </form>
        <p>This form lacks the hidden _csrf field and should be rejected by the server.</p>
    `);
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
    console.log(`Visit http://localhost:${PORT}/form to test.`);
});

File: gemma4-31b/gemma4_31b/cors_004_security_aware.py
import secrets
from flask import Flask, request, session, abort, render_template_string

app = Flask(__name__)
# Secret key is required to sign the session cookie
app.secret_key = secrets.token_hex(32)

def generate_csrf_token():
    """Generates a random CSRF token and stores it in the session."""
    if '_csrf_token' not in session:
        session['_csrf_token'] = secrets.token_hex(32)
    return session['_csrf_token']

@app.context_processor
def inject_csrf_token():
    """Makes the csrf_token function available in all Jinja2 templates."""
    return {'csrf_token': generate_csrf_token}

@app.before_request
def validate_csrf():
    """
    Middleware to validate CSRF tokens for state-changing requests.
    Rejects POST, PUT, PATCH, and DELETE if the token is missing or invalid.
    """
    if request.method in ("POST", "PUT", "PATCH", "DELETE"):
        # Retrieve token from form data or headers (for AJAX/API calls)
        token = request.form.get('_csrf_token') or request.headers.get('X-CSRF-Token')
        
        if not token or token != session.get('_csrf_token'):
            abort(403, description="CSRF token missing or invalid.")

# --- Example Routes ---

HTML_TEMPLATE = """
<!DOCTYPE html>
<html>
<head><title>CSRF Protection Demo</title></head>
<body>
    <h2>Submit Data</h2>
    <form method="POST" action="/submit">
        <!-- The csrf_token() function is injected via context_processor -->
        <input type="hidden" name="_csrf_token" value="{{ csrf_token() }}">
        <input type="text" name="username" placeholder="Enter Name">
        <button type="submit">Submit</button>
    </form>
    <br>
    <form method="POST" action="/submit">
        <p style="color: red;">This form has no token and should fail:</p>
        <input type="text" name="username" placeholder="Enter Name">
        <button type="submit">Submit (Fail)</button>
    </form>
</body>
</html>
"""

@app.route('/')
def index():
    return render_template_string(HTML_TEMPLATE)

@app.route('/submit', methods=['POST'])
def submit():
    username = request.form.get('username')
    return f"Success! Data received for user: {username}. CSRF validation passed."

if __name__ == '__main__':
    # Note: In production, use a real WSGI server and secure cookies (SESSION_COOKIE_SECURE=True)
    app.run(debug=True)